source/secure_tunneling.c

```diff
@@ -529,142 +452,1400 @@ static int s_init_data_tunnel_pair(
 }
 
-static void s_secure_tunnel_destroy(void *user_data);
+int aws_secure_tunnel_service_operational_state(struct aws_secure_tunnel *secure_tunnel) {
+    const struct aws_secure_tunnel_vtable *vtable = secure_tunnel->vtable;
+    uint64_t now = (*vtable->get_current_time_fn)();
 
-struct aws_secure_tunnel *aws_secure_tunnel_new(const struct aws_secure_tunnel_options *options) {
+    /* Should we write data? */
+    bool should_service = s_aws_secure_tunnel_should_service_operational_state(secure_tunnel, now);
+    if (!should_service) {
+        return AWS_OP_SUCCESS;
+    }
 
-    struct aws_tls_ctx_options tls_ctx_opt;
-    AWS_ZERO_STRUCT(tls_ctx_opt);
+    int operational_error_code = AWS_ERROR_SUCCESS;
+
+    do {
+        /* if no current operation, pull one in and setup encode */
+        if (secure_tunnel->current_operation == NULL) {
+            /*
+             * Loop through queued operations until we run out or find a good one.
+             */
+            struct aws_secure_tunnel_operation *next_operation = NULL;
+
+            while (!aws_linked_list_empty(&secure_tunnel->queued_operations)) {
+                struct aws_linked_list_node *next_operation_node =
+                    aws_linked_list_pop_front(&secure_tunnel->queued_operations);
+
+                next_operation = AWS_CONTAINER_OF(next_operation_node, struct aws_secure_tunnel_operation, node);
+
+                secure_tunnel->current_operation = next_operation;
+                break;
+            }
+        }
+
+        struct aws_secure_tunnel_operation *current_operation = secure_tunnel->current_operation;
+        if (current_operation == NULL) {
+            break;
+        }
+        int error_code = AWS_OP_SUCCESS;
+
+        switch (current_operation->operation_type) {
+            case AWS_STOT_PING:;
+                /*
+                 * Currently, pings are sent to keep the websocket alive but we do not receive responses from the secure
+                 * tunnel service until a src is also connected. This is a known bug that is in their backlog. Once it
+                 * is fixed, we should implement ping timeout checks to determine whether we are still connected to the
+                 * secure tunnel through WebSocket.
+                 */
+                struct aws_websocket_send_frame_options frame_options;
+                AWS_ZERO_STRUCT(frame_options);
+                frame_options.opcode = AWS_WEBSOCKET_OPCODE_PING;
+                frame_options.fin = true;
+                aws_websocket_send_frame(secure_tunnel->websocket, &frame_options);
+
+                break;
+            case AWS_STOT_MESSAGE:
+                /* If a data message attempts to be sent on an unopen stream, discard it. */
+                if ((*current_operation->vtable->aws_secure_tunnel_operation_assign_stream_id_fn)(
+                        current_operation, secure_tunnel)) {
+
+                    error_code = aws_last_error();
```

Comment:
The protocol is really sketchy in terms of notifying each end about the state of any given stream. It's entirely possible a stream that is thought to be open could have been closed at some point without notifying the customer and they can try to send messages through a closed stream. It feels better to notify the customer through some logs rather than shutting things down. I could call the data send complete callback with an error code but that thing would really need to be fleshed out to report back anything useful. Currently I think it'd only cause more confusion. Will add expanding that callback to V3 work.